csharp/src/Ice/CurrentExtensions.cs

```diff
@@ -113,25 +113,29 @@ public static OutgoingResponse createOutgoingResponse(this Current current, bool
     /// <param name="current">The current object of the corresponding incoming request.</param>
     /// <param name="exception">The exception to marshal into the response payload.</param>
     /// <returns>A new outgoing response.</returns>
-    public static OutgoingResponse createOutgoingResponse(this Current current, System.Exception exception)
+    public static OutgoingResponse createOutgoingResponse(
+        this Current current,
+        System.Exception exception,
+        Communicator? communicator = null)
     {
+        communicator ??= current.adapter.getCommunicator() ?? throw new ArgumentNullException(nameof(communicator));
         try
         {
-            return createOutgoingResponseCore(exception);
+            return createOutgoingResponseCore(exception, communicator);
         }
         catch (System.Exception ex)
         {
             // Try a second time with the marshal exception. This should not fail.
-            return createOutgoingResponseCore(ex);
+            return createOutgoingResponseCore(ex, communicator);
         }
 
-        OutgoingResponse createOutgoingResponseCore(System.Exception exc)
+        OutgoingResponse createOutgoingResponseCore(System.Exception exc, Communicator communicator)
         {
             OutputStream ostr;
 
             if (current.requestId != 0)
             {
-                ostr = new OutputStream(current.adapter.getCommunicator(), Util.currentProtocolEncoding);
+                ostr = new OutputStream(communicator, Util.currentProtocolEncoding);
```

Comment:
As I understanding it, the core issue is this call in the very uncommon situation where we receive a request over a connection without an associated object adapter.

I think the best solution is to create an OutputStream without a communicator/instance - use the parameterless OutputStream ctor.

The main difficulty is setting the encoding. Util.currentProtocolEncoding = 1.0 while the parameterless ctor sets it to Util.currentEncoding to 1.1.

Reply:
On a related note, I don't see why we keep communicator (instance) as an OutputStream field. It seems only to hand it off to other calls via the "public" instance() method.

Reply:
> The main difficulty is setting the encoding. Util.currentProtocolEncoding = 1.0 while the parameterless ctor sets it to Util.currentEncoding to 1.1.

Can we get the encoding from the request input stream?

Reply:
I would just add an encoding parameter to the currently parameterless constructor. Then, in a follow-up PR, look into eliminating OutputStream._instance.

Reply:
This doesn't work well for the collocated request handler. 

https://github.com/zeroc-ice/ice/blob/ef3c828b3a23523b022bb7632958193251fc3b5f/csharp/src/Ice/Internal/CollocatedRequestHandler.cs#L255-L259

`instance()` would be null here.